src/mlpack/methods/ne/genome.hpp

```diff
@@ -0,0 +1,334 @@
+/**
+ * @file genome.hpp
+ * @author Bang Liu
+ *
+ * Definition of the Genome class.
+ */
+#ifndef MLPACK_METHODS_NE_GENOME_HPP
+#define MLPACK_METHODS_NE_GENOME_HPP
+
+#include <cstddef>
+#include <cassert>
+#include <map>
+
+#include <mlpack/core.hpp>
+
+#include "link_gene.hpp"
+#include "neuron_gene.hpp"
+#include "utils.hpp"
+
+namespace mlpack {
+namespace ne {
+
+/**
+ * This class defines a genome.
+ # A genome is consist of a group of genes.
+ */
+class Genome {
+ public:
+  // Neurons.
+  std::vector<NeuronGene> aNeuronGenes;
+
+  // Links.
+  std::vector<LinkGene> aLinkGenes;
+
+  // Default constructor.
+  Genome() {
+    aId = -1;
+    aNumInput = 0;
+    aNumOutput = 0;
+    aFitness = DBL_MAX;
+  }
+  
+  // Parametric constructor.
+  Genome(int id,
+  	     const std::vector<NeuronGene>& neuronGenes,
+         const std::vector<LinkGene>& linkGenes,
+         int numInput,
+         int numOutput,
+         double fitness):
+    aId(id),
+    aNeuronGenes(neuronGenes),
+    aLinkGenes(linkGenes),
+    aNumInput(numInput),
+    aNumOutput(numOutput),
+    aFitness(fitness)
+  {}
+
+  // Copy constructor.
+  Genome(const Genome& genome) {
+    aId = genome.aId;
+    aNeuronGenes = genome.aNeuronGenes;
+    aLinkGenes = genome.aLinkGenes;
+    aNumInput = genome.aNumInput;
+    aNumOutput = genome.aNumOutput;
+    aFitness = genome.aFitness;
+  }
+
+  // Destructor.
+  ~Genome() {}
+
+  // Operator =.
+  Genome& operator =(const Genome& genome) {
+    if (this != &genome) {
+      aId = genome.aId;
+      aNeuronGenes = genome.aNeuronGenes;
+      aLinkGenes = genome.aLinkGenes;
+      aNumInput = genome.aNumInput;
+      aNumOutput = genome.aNumOutput;
+      aFitness = genome.aFitness;
+    }
+
+    return *this;
+  }
+
+  // Get genome id.
+  int Id() const { return aId; }
+
+  // Set genome id.
+  void Id(int id) { aId = id; }
+
+  // Get input length.
+  int NumInput() { return aNumInput; }
+
+  // Set input length.
+  void NumInput(int numInput) { aNumInput = numInput; }
+
+  // Get output length.
+  int NumOutput() { return aNumOutput; }
+
+  // Set output length.
+  void NumOutput(int numOutput) { aNumOutput = numOutput; }
+
+  // Set fitness.
+  void Fitness(double fitness) { aFitness = fitness; }
+
+  // Get fitness.
+  double Fitness() const { return aFitness; }
+
+  // Get neuron number.
+  int NumNeuron() const {
+    return aNeuronGenes.size();
+  }
+  
+  // Get link number.
+  int NumLink() const {
+    return aLinkGenes.size();
+  }
+
+  // Get input length.
+  int GetNumInput() {
```

Comment:
We already have NumInput() which is used to return the value of aNumInput.
This GetNumInput() is get the number of numInput by loop, rather than look at the aNumInput. I implement this function as I found that previously the real numInput sometimes are not the same with aNumInput.